Replace old version command with brigadier equivalent

NonSwag opened the pull request:
Not sure what to do about the command description, I would appreciate some input on that.

NonSwag committed:
remove old version command

NonSwag committed:
replaced ambiguous with direct imports

NonSwag committed:
add brigadier `version` command

NonSwag committed:
register new version command

NonSwag committed:
move command registration from Commands to ReloadableServerResources

Doc94 commented on paper-api/src/main/java/org/bukkit/command/SimpleCommandMap.java:
```diff
@@ -36,7 +33,7 @@ public SimpleCommandMap(@NotNull final Server server, Map<String, Command> backi
 
     private void setDefaultCommands() {
-        register("bukkit", new VersionCommand("version"));
+        //register("bukkit", new VersionCommand("version")); // Paper
```
Not better just remove?

NonSwag replied:
Wasn't quite sure because the /plugin command is commented out as well
didn't want to remove it immediately

Doc94 commented on paper-server/src/main/java/io/papermc/paper/command/VersionCommand.java:
```diff
@@ -0,0 +1,183 @@
+                .findAny())
+            .ifPresentOrElse(plugin -> describeToSender(plugin, sender), () -> {
+                sender.sendMessage(Component.text("This server is not running any plugin by that name."));
+                sender.sendMessage(Component.text("Use /plugins to get a list of plugins."));
```
Can be good allow click the plugins text?

NonSwag replied:
I thought of that too but wanted to keep it in pair with the current behavior
but if others are in favor of this too, why not :+1:

Doc94 replied:
Can be good make improvements like was the plugins command

Malfrador replied:
iirc that was removed again? At least the hover message was, because it made the resulting message too large to be sent to the client.

NonSwag replied:
I highly doubt that, chat messages can be way longer

Doc94 commented on paper-server/src/main/java/io/papermc/paper/command/VersionCommand.java:
```diff
@@ -0,0 +1,183 @@
+        return Command.SINGLE_SUCCESS;
+    }
+
+    private static void describeToSender(@NotNull Plugin plugin, @NotNull CommandSender sender) {
```
NotNull are not necesary if afe using the NullMarked

NonSwag replied:
copy paste my beloved xD

NonSwag committed:
Deprecated `VersionFetcher#getVersionMessage(String)`

Marked the method as `@Deprecated` and provided a default implementation redirecting to the new `getVersionMessage()` method.

NonSwag committed:
[ci skip] remove comments

NonSwag committed:
[ci skip] remove redundant `@NotNull` annotations

NonSwag committed:
Suggest /plugins when clicking on "unknown plugin" message

NonSwag committed:
little cleanup

NonSwag committed:
[ci skip] renamed method

Doc94 commented on paper-server/src/main/java/io/papermc/paper/command/VersionCommand.java:
```diff
@@ -0,0 +1,160 @@
+import org.jspecify.annotations.Nullable;
+
+@NullMarked
+public class VersionCommand {
```
Maybe this can be named "PaperVersionCommand" based in the rest of classes

Doc94 commented on paper-server/src/main/java/io/papermc/paper/command/VersionCommand.java:
```diff
@@ -0,0 +1,160 @@
+    private static boolean versionTaskStarted = false;
+    private static long lastCheck = 0;
+
+    public static void register(CommandDispatcher<CommandSourceStack> dispatcher) {
```
Why this method for register commands? the rest of classes just depends on PaperCommands unless in future the team maybe wanna move to brigadier...

Owen1212055 reviewed (changes requested):
I think we should instead see about creating an internal PaperLifecycleEventManager somewhere and instead registering these commands on start.

We'd have some kind of paper LifecycleEventOwner.

This way you can use the api to implement everything, and register it through the api allowing you to add descriptions.

NonSwag committed:
Merge branch 'main' into brig-version-command

# Conflicts:
#	paper-server/patches/sources/net/minecraft/server/ReloadableServerResources.java.patch

NonSwag committed:
renamed to PaperVersionCommand

NonSwag committed:
register paper version command

NonSwag committed:
use paper brig

NonSwag committed:
add description constant

NonSwag committed:
fix namespace

Owen1212055 committed:
Refactor

Owen1212055 committed:
Redo exception handling

Owen1212055 committed:
Oops

Owen1212055 committed:
Oops

Owen1212055 committed:
Use when complete

Owen1212055 committed:
Rework to not compose

Owen1212055 committed:
Add oxford comma

Owen1212055 committed:
Ok final

Owen1212055 committed:
Improve on Copilot's comment

Owen1212055 commented on paper-server/src/main/java/io/papermc/paper/command/PaperVersionCommand.java:
```diff
@@ -0,0 +1,183 @@
+        Plugin plugin = Bukkit.getPluginManager().getPlugin(pluginName);
+        if (plugin == null) {
+            plugin = Arrays.stream(Bukkit.getPluginManager().getPlugins())
+                .filter(checkPlugin -> checkPlugin.getName().toLowerCase(Locale.ROOT).contains(pluginName))
```
Shut up Meg

kennytv commented on paper-server/src/main/java/io/papermc/paper/command/PaperVersionCommand.java:
```diff
@@ -0,0 +1,183 @@
+    private CompletableFuture<Suggestions> suggestPlugins(final CommandContext<CommandSourceStack> context, final SuggestionsBuilder builder) {
+        for (final Plugin plugin : Bukkit.getPluginManager().getPlugins()) {
+            final String name = plugin.getName();
+            if (StringUtil.startsWithIgnoreCase(name, builder.getRemainingLowerCase())) {
```
No need to getRemainingLowerCase, else put it into an extra variable outside the loop

kennytv commented on paper-server/src/main/java/io/papermc/paper/command/PaperVersionCommand.java:
```diff
@@ -0,0 +1,183 @@
+                sender.sendMessage(Component.text("Could not fetch version information!", NamedTextColor.RED));
+                MinecraftServer.LOGGER.warn("Could not fetch version information!", throwable);
+            }
+        }, MinecraftServer.getServer());
```
this doesn't have to be thrown to main

Owen1212055 committed:
Cache + complete on whatever thread we do not care

Owen1212055 committed:
Small update
